flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableFactory.java

```diff
@@ -0,0 +1,121 @@
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.addons.hbase.HBaseValidator.CONNECTOR_HBASE_TABLE_NAME;
+import static org.apache.flink.addons.hbase.HBaseValidator.CONNECTOR_HBASE_ZK_QUORUM;
+import static org.apache.flink.addons.hbase.HBaseValidator.CONNECTOR_TYPE_VALUE_HBASE;
+import static org.apache.flink.addons.hbase.HBaseValidator.CONNECTOR_VERSION_VALUE_143;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_PROPERTY_VERSION;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_TYPE;
+import static org.apache.flink.table.descriptors.ConnectorDescriptorValidator.CONNECTOR_VERSION;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_NAME;
+import static org.apache.flink.table.descriptors.Schema.SCHEMA_TYPE;
+
+/**
+ * Factory for creating configured instances of {@link HBaseTableSource} or sink.
+ */
+public class HBaseTableFactory implements StreamTableSourceFactory<Row> {
+
+	@Override
+	public StreamTableSource<Row> createStreamTableSource(Map<String, String> properties) {
+		final DescriptorProperties descriptorProperties = getValidatedProperties(properties);
+		// create default configuration from current runtime env (`hbase-site.xml` in classpath) first,
+		Configuration hbaseClientConf = HBaseConfiguration.create();
+		String hbaseZk = properties.get(CONNECTOR_HBASE_ZK_QUORUM);
+		hbaseClientConf.set(HConstants.ZOOKEEPER_QUORUM, hbaseZk);
+		String hTableName = descriptorProperties.getString(CONNECTOR_HBASE_TABLE_NAME);
+		TableSchema tableSchema = descriptorProperties.getTableSchema(SCHEMA);
+		HBaseTableSchema hbaseSchema = validateTableSchema(tableSchema);
+		return new HBaseTableSource(hbaseClientConf, hTableName, hbaseSchema, null);
+	}
+
+	private HBaseTableSchema validateTableSchema(TableSchema schema) {
+		HBaseTableSchema hbaseSchema = new HBaseTableSchema();
+		String[] fieldNames = schema.getFieldNames();
+		TypeInformation[] fieldTypes = schema.getFieldTypes();
+		for (int i = 0; i < fieldNames.length; i++) {
+			String name = fieldNames[i];
+			TypeInformation<?> type = fieldTypes[i];
+			if (type instanceof RowTypeInfo) {
+				RowTypeInfo familyType = (RowTypeInfo) type;
+				String[] qualifierNames = familyType.getFieldNames();
+				TypeInformation[] qualifierTypes = familyType.getFieldTypes();
+				for (int j = 0; j < familyType.getArity(); j++) {
+					hbaseSchema.addColumn(name, qualifierNames[j], qualifierTypes[j].getTypeClass());
+				}
+			} else {
+				hbaseSchema.setRowKey(name, type.getTypeClass());
+			}
+		}
+		return hbaseSchema;
+	}
+
+	private DescriptorProperties getValidatedProperties(Map<String, String> properties) {
+		final DescriptorProperties descriptorProperties = new DescriptorProperties(true);
+		descriptorProperties.putProperties(properties);
+		new HBaseValidator().validate(descriptorProperties);
```

Comment:
Could we introduce a singleton instance inside `HBaseValidator` instead of create a new one in each invocation?

Reply:
I think it's a good idea. However, all the existing validator keeps this style, for example: KafkaValidator, CsvValidator, etc... So I prefer to follow this for now.

Reply:
You're right, unfortunately... Let's find some time to clean this up, in another PR.